prereise/gather/griddata/hifld/data_access/load.py

```diff
@@ -186,6 +189,104 @@ def get_hifld_electric_substations(path):
+            coordinate pairs.
+        :return: (*networkx.Graph*) -- a minimum spanning tree from the segments.
+        :raises ValueError: if the minimum spanning tree is not linear
+        """
+        # Build a Graph representing the existing connections between segment endpoints
+        g = nx.Graph()
+        # add a 'start' and 'end' node for each segment, and a zero-cost edge b/w them
+        for i in range(len(segments)):
+            (lon1, lat1), (lon2, lat2) = segments[i][0], segments[i][-1]
+            segment_angular_distance = angular_distance(
+                latlon_to_xyz(lat1, lon1), latlon_to_xyz(lat2, lon2)
+            )
+            if segment_angular_distance < threshold:
+                # Segment is too small, so we won't add this to the graph
+                continue
+            g.add_nodes_from([(i, "start"), (i, "end")])
+            g.add_edge((i, "start"), (i, "end"), weight=0)
+        # Add edge weights by the angular distance between different segment endpoints
+        for (seg1, which1), (seg2, which2) in itertools.combinations(g.nodes, 2):
+            if seg1 == seg2:
+                # no edge needed, since there's already a zero-cost edge there
+                continue
+            # Extract latitude and longitude from the appropriate end of each segment
+            lon1, lat1 = segments[seg1][0 if which1 == "start" else -1]
+            lon2, lat2 = segments[seg2][0 if which2 == "start" else -1]
+            edge_angle = angular_distance(
+                latlon_to_xyz(lat1, lon1), latlon_to_xyz(lat2, lon2)
+            )
+            g.add_edge((seg1, which1), (seg2, which2), weight=edge_angle)
+        # Identify edges to connect the segments
+        tree = nx.algorithms.tree.minimum_spanning_tree(g)
+        # Ensure that the minimum spanning tree is linear
+        if max(dict(nx.degree(tree)).values()) > 2:
+            raise ValueError("The minimum spanning tree isn't linear")
+        return tree
+
+    # Sort the segments by length, in case the minimum spanning tree isn't linear
+    sorted_segments = sorted(
+        segments,
+        key=lambda x: angular_distance(
+            latlon_to_xyz(x[0][1], x[0][0]),  # first point
+            latlon_to_xyz(x[-1][1], x[-1][0]),  # last point
+        ),
+    )
+    # Make a minimum spanning tree, dropping small segments as necessary for linearity
+    for i in range(len(segments)):
+        try:
+            linear_segments = sorted_segments[i:] if i > 0 else sorted_segments
+            tree = _generate_linear_spanning_tree(linear_segments)
+            # We found a good minimum spanning tree, so we can exit
+            break
+        except ValueError:
+            # The minimum spanning tree wasn't good, go to the next iteration
+            pass
+
+    if len(segments) == 1:
+        # There's only one segment in the list, so no joining is necessary
+        return segments[0]
```

Comment:
Should this condition go at the top, right after the docstring of `_join_line_segments`?

Reply:
the condition is on `segments` (the input parameter) not on `linear_segments`

Reply:
Ah, yeah, good catch. I think it should be on `linear_segments`.

Reply:
Single-segment lines are the vast majority in the latest version of the dataset, ~99.9%. Multi-segment lines are the exceptions.

Reply:
I guess we can speed up the code by avoiding going into `_generate_linear_spanning_tree` if we don't have to.

Reply:
Okay, this is updated in the latest commit. I added some tests as well that check whether the returned object `is` the entry from the original list, to ensure that we're not entering the tree-building and joining logic when we don't have to, plus a couple other tests for what happens when we do enter the tree-building/joining logic.

Reply:
Thanks